### PR TITLE
Leverage mypy to check str.format expressions

### LINKED ISSUE #42: Support attribute references of positional/keyword arguments with string formatting
```python
from loguru import logger

class Foo:
  bar = "baz"

foo = Foo()

# Both the following are valid syntax, but mypy flags them
logger.debug('The bar is "{0.bar}"', foo)  # note: Expected 0 but found 1 positional arguments for log message
logger.debug('The bar is "{my_foo.bar}"', my_foo=foo)  # error: my_foo.bar keyword argument is missing
```

```fish
❯ pip freeze | grep loguru
loguru==0.5.2
loguru-mypy==0.0.2
```

### PR DESCRIPTION
**Pull request by ThibaultLemaire**
```py
from loguru import logger

class Foo:
bar = "baz"

foo = Foo()

logger.debug('The bar is "{my_foo.bar}"', my_foo=foo)  # ✓
logger.debug('The bar is "{my_foo.bar}"', my_foo="not foo")  # error: "str" has no attribute "bar"
```

For the record, this has been developed and tested against `mypy==0.782`.

Closes #42

### CODE REVIEW

File: loguru_mypy/__init__.py [hunk 1]
```diff
@@ -73,6 +75,14 @@ def _loguru_logger_call_handler(
     except IndexError:
         call_kwargs = {}
 
+    call_expr = CallExpr(
+        callee=MemberExpr(expr=log_msg_expr, name="format"),
+        args=call_args,
+        arg_kinds=ctx.arg_kinds[1],
+        arg_names=ctx.arg_names[1],
+    )
+    ctx.api.expr_checker.visit_call_expr(call_expr)
```
Comment (ThibaultLemaire) [inline loguru_mypy/__init__.py:84]: Now, I know the fact that `ctx.api` is an instance of `TypeChecker` is just a implementation detail, and this isn't part of mypy's plugin API that you can access the `expr_checker` there (and mypy is not happy with that as the `CheckerPluginInterface` doesn't expose that, but is really quite convenient to just be able to hook into mypy's existing methods instead reinventing the wheel and parsing this all by ourselves.

File: loguru_mypy/__init__.py [hunk 2]
```diff
@@ -73,6 +75,14 @@ def _loguru_logger_call_handler(
     except IndexError:
         call_kwargs = {}
 
+    call_expr = CallExpr(
+        callee=MemberExpr(expr=log_msg_expr, name="format"),
+        args=call_args,
+        arg_kinds=ctx.arg_kinds[1],
+        arg_names=ctx.arg_names[1],
```
Comment (ThibaultLemaire) [inline loguru_mypy/__init__.py:82]: I'm not sure I understand mypy's structure there. Why are all those `arg_*` attributes lists of lists, and why is the second one the one I'm interested in? It seems it has to do with "formal arguments" but I'm not sure I know what that means.

**Comment by ThibaultLemaire** [comment]
@kornicameister How do you track which versions of mypy are compatible? Since I am tapping into mypy's internals I am worried it will break with upcoming versions, so I think we should have a way of making that visible.

File: loguru_mypy/__init__.py (hunk 1 above)
Comment (kornicameister) [reply loguru_mypy/__init__.py]: I am pretty sure I have stumbled onto these on my own several times. I do not mind that as long as we keep a close eye on that. I am pretty sure I need to finally find time and start testing against several latest releases of mypy + pin the latest on plugin hook.

Comment (kornicameister) [reply loguru_mypy/__init__.py]: https://github.com/kornicameister/loguru-mypy/issues/47

File: loguru_mypy/__init__.py (hunk 2 above)
Comment (kornicameister) [reply loguru_mypy/__init__.py]: Neither do I, TBH I debugged the hell out of it when implementing for the first time. 
From what I remember, it has something to do with the fact that mypy does not use enums, instead they have integers. So the second is is basically about either positional or named arguments (can't remember which one is which). Anyway, I am pretty sure there is a constant somewhere saying that `1=KWARGS` and `0=ARGS` or vice-versa.

Does that help at all?

**Comment by kornicameister** [comment]
@ThibaultLemaire thx for pull request. It looks really nice, I'd like to hear more about that `visit_expr` you came up with and how did you that. `mypy` is magical and that's a magic I am seeing :D.

Anyway,I really like your job on error messages although I wonder if having them more accurate wasn't better. Could you elaborate why did you change those?

PS. sorry for being so lazy about reviewing, I had my evenings filed with either micro-management on 2 children or overtime from work :/

File: loguru_mypy/__init__.py (hunk 1 above)
Comment (kornicameister) [reply loguru_mypy/__init__.py]: I think that answers your question below about `mypy` version.

File: loguru_mypy/__init__.py [hunk 3]
```diff
@@ -47,6 +50,35 @@ class Opts(t.NamedTuple):
...
+        arg_names=ctx.arg_names[1] + ctx.arg_names[2],
+    )
+    call_expr.set_line(log_msg_expr)
+
+    # WARNING: `ctx.api` *is* technically a `mypy.checker.TypeChecker` so the cast is
+    # safe to make, however, mypy says this should be an implementation detail.
+    # So, anything that's not part of the `CheckerPluginInterface` should be expected to
+    # change. See https://github.com/python/mypy/issues/6617
+    type_checker = t.cast(TypeChecker, ctx.api)
+    type_checker.expr_checker.visit_call_expr(call_expr)
```
Comment (kornicameister) [inline loguru_mypy/__init__.py:79]: PS. Can add a `try...except` here to catch `AttributeError`? An error we log or re-throw (up to you) should basically say that someone might be using incompatible mypy version.

File: typesafety/test_str_format.yml [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+    - level: exception
+  main: |
+    from loguru import logger
+
+    class Foo:
+      bar = "baz"
+
+    foo = Foo()
+
+    logger.{{ level }}('The bar is "{0.bar}"', foo)
```
Comment (kornicameister) [inline typesafety/test_str_format.yml:18]: @Delgan that is one case I'd not dream of :D thx @ThibaultLemaire  for spotting that.

File: typesafety/test_str_format.yml [hunk 5]
```diff
@@ -0,0 +1,33 @@
+---
+- case: ok__dot_access
```
Comment (kornicameister) [inline typesafety/test_str_format.yml:2]: I wonder if that file shouldn't be called `test_dot_access.yml` or `test_objects_access.yml`. From the looks of it you are trying to verify just that, right?

File: typesafety/test_builtin_severities.yml [hunk 6]
```diff
@@ -60,11 +60,11 @@
     logger.error('test {} {} {}', 11, 22, 33, 44)
     logger.exception('{}', 1, 2)
   out: |
-    main:3: note: Expected 0 but found 1 positional arguments for log message
-    main:4: note: Expected 1 but found 3 positional arguments for log message
-    main:5: note: Expected 2 but found 3 positional arguments for log message
-    main:6: note: Expected 3 but found 4 positional arguments for log message
-    main:7: note: Expected 1 but found 2 positional arguments for log message
+    main:3: error: Not all arguments converted during string formatting  [str-format]
```
Comment (kornicameister) [inline typesafety/test_builtin_severities.yml:63]: hmm...now looking at it it seems you've managed to take `mypy` to verify formatting on its own, correct? 
If that's the case I think I am taking back what I wrote about changing error messages.
If that's the case we will have `loguru-mypy` aligned to what users of `mypy` experience with string interpolation, which might be desirable.

**Comment by Delgan** [comment]
Well, I probably don't understand half of the magic of `mypy`, but the result looks great! Thanks for your work. :)

**Comment by ThibaultLemaire** [comment]
Hey, looks like it's my turn to apologize for the delay, last week was crazy at work.

Thank you Delgan, I have been focusing precisely on keeping the magic on mypy's side with this PR and not having to understand it :wink: Hopefully we can make that work in satisfying way

File: typesafety/test_builtin_severities.yml (hunk 6 above)
Comment (ThibaultLemaire) [reply typesafety/test_builtin_severities.yml:63]: Replying to the following as well:

>  I'd like to hear more about that visit_expr you came up with and how did you that.
>
>  it seems you've managed to take mypy to verify formatting on its own, correct?

Yes that's exactly it. All I do is build a new syntax tree from the existing one, feed it back to mypy, and let it do its thing. Effectively turning

```python
logger.debug('The bar is "{my_foo.bar}"', my_foo=foo)
```

into

```python
logger.debug('The bar is "{my_foo.bar}"'.format(my_foo=foo))
```

to the eyes of mypy.

---

> I really like your job on error messages although I wonder if having them more accurate wasn't better. Could you elaborate why did you change those?
>
>  I think I am taking back what I wrote about changing error messages.
[...] we will have loguru-mypy aligned to what users of mypy experience with string interpolation, which might be desirable.

I agree we are losing a bit of information there, and I'm surprised mypy does not give much to begin with. I was a bit hesitant at first, but if you're cool with it, I'm cool with it :)

File: loguru_mypy/__init__.py (hunk 2 above)
Comment (ThibaultLemaire) [reply loguru_mypy/__init__.py]: Yeah I figured that up along the way :slightly_smiling_face:

File: typesafety/test_str_format.yml (hunk 5 above)
Comment (ThibaultLemaire) [reply typesafety/test_str_format.yml:2]: Fine by me, I honestly had no idea how to name that file

File: loguru_mypy/__init__.py (hunk 3 above)
Comment (ThibaultLemaire) [reply loguru_mypy/__init__.py:79]: Good idea, but an exception would bubble up at parse time and completely crash mypy... On the other hand, a mypy error message might look weird to the user (as it's not really a problem with their code) but would let mypy continue on with the rest of the analysis.

Also, how can we test that code path? I don't know of a version of mypy that would trigger that

I'll go with the mypy message message for now, tell me what you think.

Comment (kornicameister) [reply loguru_mypy/__init__.py]: Version of mypy is related but external topic. I've already created an issue for that which also I am happy to accept PR.
I am afraid that with ongoing craziness (also work) I might just lose it out of focus :(. Anyway...it seems like entire world gos crazy nowadays, so yeah :D

**Comment by kornicameister** [comment]
@ThibaultLemaire  would you mind checking #31 ? Would you change invalid that PR?

**Comment by ThibaultLemaire** [comment]
Hm, I can imagine why my work would conflict this PR, but I'm not sure how exactly the two would interact.

Also, what exactly do you mean by "invalid[ate]" ? If you mean make obsolete, I don't think so: my changes have no way of knowing if `record=True` has been passed, but if you mean conflict in a way that would require to update #31, I'm afraid so, yes.

I only glanced at it though, so I'll take the time to investigate properly.

**Comment by ThibaultLemaire** [comment]
Okay, so, after further investigation, I reckon you could adapt your work on #31 to make it work side by side, but then you'd have a bit of an overlap.

To take your example code from #18, and compare the (theoretical) results of #31 and #43 (this PR):

```python
logger.info('{record["msg"]}') # loguru -> ❌, #31 -> ❌, #43 -> ❌
logger.opt(record=True).info('{record["msg"]}') # loguru -> ✔, #31 -> ✔, #43 -> ❌
```

As you can see, there would be duplicate error messages, as well as contradicting messages where my work has shortcomings and can't understand that `record` is a valid keyword argument.

The issue is that I rely entirely on mypy's internals and have no control over them, so I couldn't silence that error even if I wanted to.

What we could do, though, -- and, to me, that would be the best solution -- would be to pass the `record` argument to mypy to validate it as well. In the same spirit to what I've done here, this would mean making mypy validate

```python
logger.opt(record=True).info('{record["msg"]}')
```

as if it were something like

```python
logger.info('{record["msg"]}'.format(record={}))
```

Unfortunately, I don't know how much of your work we could reuse if we go down that path.

**Comment by ThibaultLemaire** [comment]
Also related to that, I have noticed that you store the loggers the plugin has previously encountered in the `self._known_loggers` dict, to deal with `.opt()` configuration, but this is discouraged by [the mypy documentation](https://mypy.readthedocs.io/en/stable/extending_mypy.html#high-level-overview) (emphasis mine):

> Plugin developers should ensure that their plugins work well in incremental and daemon modes. In particular, **plugins should not hold global state** due to caching of plugin hook results.

I wonder if we could rather play on the type we return from the plugin to store that configuration, which should save us a lot of trouble in the future as well.

To illustrate, I was wondering if we could make things work a bit like this:

```python
from loguru import logger  # type is inferred by mypy to be Logger

with_record: LoggerWithRecord = logger.opt(record=True) 
# Those would be custom types that we return, in order to know what to do next time we encounter it
lazy: LazyLogger = logger.opt(lazy=True)
with_ip: BoundLogger = logger.bind(ip="192.168.0.1")
with_ip_in_record: BoundLoggerWithRecord = with_ip.opt(record=True)

# And here mypy should pass us all the info we need to validate it properly
with_ip_in_record.info("Your local ip is {record["ip"]}")
```

**Comment by kornicameister** [comment]
@ThibaultLemaire for the global state, I was not aware of that. If could drop the issue for that matter, I'd appreciate. Just put a link to your comment.

As for the #31 I think I'd prefer to have mypy deal with string formatting if possible. I imagine them doing far better job.
This is rather interesting glitch, I certainly need time to even get my head around. I mean, don't get it wrong, your idea is far nicer compared to mine but this `record` thingy makes it hard to just accept. 
As you fairly pointed out it simply won't work now.

Unless, we can get our hands on `record=True` and somehow push into kwarg arguments of formatting call a type containing all fields from a record `loguru` defines. Anyway, `record=True` is not a point of interest on this PR. It is for #18 to resolve.
I will play around with your solution and I believe in the end of the week, unless something pops out, we can merge that.

@Delgan do you have any interesting cases we could use to challenge this PR? If you do, please put a comment :)

**Comment by Delgan** [comment]
I tried hard but could not generate a error specific to the `_check_str_format_call` function. Congrats for mastering `mypy` internals!

However,  some tests led me to open #49. :p

**Comment by kornicameister** [comment]
@thx @Delgan , I will try to look at this if possible..or rather if I find time...it is horrible how little of that I have :D

**Review by kornicameister — APPROVED** [review]
Ok, lets proceed with it.
As I said before, for me it is better to reuse mypy because it might help with both making it work more predictable and more compliant with different mypy versions.
Also, given that @Delgan did not exactly raise incompatibility between `logoru` and this PR on runtime, I don't see anything that would prevent us from doing so.
Errors reported earlier are actually related to us using `StrExpr` all the time.

**Comment by kornicameister** [comment]
@ThibaultLemaire I am merging this. Thanks a lot for patience and waiting for me here :)

**Pull request merged**